---
title: "Code for the analyses of a Response article to the paper: "Therapy-induced APOBEC3A drives evolution of persistent cancer cells"
author: "Orestis Nousias"
date: "09/13/2023"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

##CODE

library(cancereffectsizeR)
library(data.table)
library(dplyr)

#Download the data from the manuscript link Supplementary Table 7 that contains the calls for the patients and cell lines: https://static-content.springer.com/esm/art%3A10.1038%2Fs41586-023-06303-1/MediaObjects/41586_2023_6303_MOESM7_ESM.txt 

# Define the URL of the file to download
file_url <- "https://static-content.springer.com/esm/art%3A10.1038%2Fs41586-023-06303-1/MediaObjects/41586_2023_6303_MOESM7_ESM.txt"

# Define the destination file name
dest_file <- "Isozaki_et_al.calls.txt"

# Use download.file to fetch the file
download.file(file_url, dest_file, method="auto")

#Process the file to rename columns and retain only the data for the 24 patients
#Rename column names: name -> Unique_Patient_Identifier,	chr -> Chromosome,	pos -> Start_Position,	REF -> Reference_Allele,	ALT -> Tumor_Allele

# Step 1: Load the file
data <- read.table("Isozaki_et_al.calls.txt", header = TRUE, sep = "\t", stringsAsFactors = FALSE, skip = 3)

# Step 2: Rename the columns
data <- rename(data, Unique_Patient_Identifier = name, Chromosome = chr, Start_Position = pos, Reference_Allele = REF, Tumor_Allele = ALT)

# Step 3: Save the new file
write.table(data, "APOBEC_response_data.txt", sep = "\t", row.names = FALSE, quote = FALSE)

#Start running cancer effect size R
apobec_maf <- preload_maf(maf = fread("APOBEC_response_data.txt", sep = "\t"),
                           refset = "ces.refset.hg19")

#Create CESAnalysis and load data

cesa <- CESAnalysis(refset = "ces.refset.hg19")

cesa <- load_maf(cesa = cesa, maf = apobec_maf, maf_name = "APOBEC")

#CESA WITH Lung signatures

nsclc_exclusions <- suggest_cosmic_signature_exclusions(cancer_type = "Lung-AdenoCA", treatment_naive = NULL)

cesa <- trinuc_mutation_rates(
    cesa = cesa,
    signature_set = ces.refset.hg19$signatures$COSMIC_v3.2,
    signature_exclusions = nsclc_exclusions
)

#CONTINUE WITH GENE RATES

#print(ces.refset.hg19$covariates)

cesa <- gene_mutation_rates(cesa, covariates = ces.refset.hg19$covariates$lung)

#Get mutation rates per variant for the patients of interest

#all 4 patients
samples_to_check <- c("MGH808-A", "MGH808-E", "MGH808-F","MGH953-1", "MGH953-2-CL", "MGH953-3-CL", "MGH953-4-CL", "MGH953-6", "MGH10032-AS-mediastinal", "MGH10032-LL-supraclavicular-LN", "MGH10032-R-hilar-LN", "MGH10032-R-paratracheal-LN", "MGH10032-RLL", "MGH10032-pre-treatment", "MGH086-0", "MGH086-00", "MGH086-000", "MGH086-1", "MGH086-2", "MGH086-3", "MGH086-4", "MGH086-5")

#Variant id's and corresponding change in the gene
#variant_id	gene_change
#2:29443610_C>T	ALK_D1203N
#2:29443589_C>T	ALK_E1210K
#2:29443613_C>T	ALK_G1202R
#2:29432682_C>G	ALK_G1269A
#2:29445469_G>C	ALK_L1122V
#2:29443631_G>T	ALK_L1196M
#2:29443600_G>C	ALK_S1206C
#7:55249092_G>C	EGFR_C797S
#7:55249071_C>T	EGFR_T790M
#2:29448369_G>A	ALK_L1044F
#2:29445213_A>T	ALK_I1171N
#2:29443625_G>A ALK_L1198F
#7:55259515_T>G EGFR_L858R

variants_to_check <- c("7:55249071_C>T", "7:55249092_G>C", "7:55259515_T>G", "2:29443613_C>T", "2:29443631_G>T", "2:29443589_C>T", "2:29432682_C>G", "2:29443610_C>T", "2:29445469_G>C", "2:29443600_G>C", "2:29443625_G>A", "2:29448369_G>A","2:29445213_A>T")

# Generate the output using the baseline_mutation_rates function
output <- baseline_mutation_rates(cesa = cesa, variant_ids = variants_to_check, samples = samples_to_check)

write.table(output, file="Mutation_rates_per_variant.tsv", sep="\t", quote=FALSE, row.names=FALSE)

# Load the data
rename <- fread("Mutation_rates_per_variant.tsv") # Replace with your actual file path

# Create the mapping from variant IDs to gene names
name_mapping <- setNames(c("ALK_D1203N", "ALK_E1210K", "ALK_G1202R", "ALK_G1269A", 
                           "ALK_L1122V", "ALK_L1196M", "ALK_S1206C", "EGFR_C797S", 
                           "EGFR_T790M", "ALK_L1044F", "ALK_I1171N","ALK_L1198F","EGFR_L858R"),
                         c("2:29443610_C>T", "2:29443589_C>T", "2:29443613_C>T", 
                           "2:29432682_C>G", "2:29445469_G>C", "2:29443631_G>T", 
                           "2:29443600_G>C", "7:55249092_G>C", "7:55249071_C>T", 
                           "2:29448369_G>A", "2:29445213_A>T","2:29443625_G>A","7:55259515_T>G"))


# Rename the columns based on the mapping
names(rename) <- sapply(names(output), function(x) ifelse(x %in% names(name_mapping), name_mapping[x], x))

# Write the modified data frame back to a TSV file
fwrite(rename, file="Mutation_rates_per_variant_named.tsv", sep="\t", quote=FALSE)

file.remove("Mutation_rates_per_variant.tsv")

_____________________________________________________________________________
_____________________________________________________________________________

#Get the signature contributions for these variants

# Assuming you have defined 'samples_to_check' and 'variants_to_check' earlier
# Extract signature data
signature_data <- cesa@trinucleotide_mutation_weights$signature_weight_table

# Merge with MAF data
merged_data <- merge(cesa@maf, signature_data, by = "Unique_Patient_Identifier")

# Filter for specific variants and sample IDs
filtered_data <- merged_data[merged_data$variant_id %in% variants_to_check & 
                             merged_data$Unique_Patient_Identifier %in% samples_to_check, ]

# Convert the list-type 'genes' column to a character column
filtered_data$genes <- sapply(filtered_data$genes, function(x) ifelse(length(x) > 0, toString(x), NA))

# Write the filtered data to a TSV file, this file contains more data than we need, but it serves as an additional reference point
write.table(filtered_data, file = "filtered_data.tsv", sep = "\t", row.names = FALSE)

# Read the TSV file into a data frame
filtered_data <- fread("filtered_data.tsv")

# Define the columns to remove
columns_to_remove <- c("Chromosome", "Start_Position", "Reference_Allele", "Tumor_Allele",
                       "variant_type", "variant_id", "genes", "top_gene", "total_snvs", 
                       "sig_extraction_snvs", "group_avg_blended")

# Remove the specified columns
filtered_data <- filtered_data[, !names(filtered_data) %in% columns_to_remove, with = FALSE]

# Remove columns that only contain 0s
filtered_data <- filtered_data[, sapply(filtered_data, function(col) !all(col == 0)), with = FALSE]

# Move 'top_consequence' column to the first position and rename it to 'variant_id'
setnames(filtered_data, old = "top_consequence", new = "variant_id")
filtered_data <- filtered_data[, c("variant_id", setdiff(names(filtered_data), "variant_id")), with = FALSE]

# Write the modified data frame back to a TSV file
fwrite(filtered_data, file="Source_probabilities.tsv", sep="\t", quote=FALSE)


#Get the mutation rates for each variant in the Source probabilities file

# Read the datasets
mutation_rates <- fread("Mutation_rates_per_variant_named.tsv")
source_probabilities <- fread("Source_probabilities.tsv")

# Melt the mutation_rates data.table from wide to long format
mutation_rates_long <- melt(mutation_rates, id.vars = "Unique_Patient_Identifier", 
                            variable.name = "variant_id", value.name = "VAR_MUTRATE")

# Merge the long format mutation_rates with source_probabilities on 'variant_id'
source_probabilities <- merge(source_probabilities, mutation_rates_long, 
                              by = c("variant_id", "Unique_Patient_Identifier"), all.x = TRUE)

# Write the updated source_probabilities to a new TSV file
fwrite(source_probabilities, file="Source_probabilities_with_mutrates.tsv", sep="\t", quote=FALSE)

_________________________________________________________________________
_________________________________________________________________________

#Hypothetical Source Probabilities

# Assuming cesa@maf is a data frame with a column 'variant_id' and cesa@maf$genes is a list of genes

#Jeff Mandel-Hypothetical_Source_Probabilities

# Input the CESAnalysis, the sample ID (Unique_Patient_Identifier), and the variant ID.
# Replace the signature definitions if you didn't use COSMIC v3.2. (at the hypotheticalSourceProbFunction.R)

#source it 
source("hypotheticalSourceProbFunction.R")


#then run this loop for these ids:

# Define the variant IDs
variant_ids <- c("2:29448369_G>A", #ALK_L1044F
                 "2:29445213_A>T", #ALK_I1171N
                 "2:29443625_G>A", #ALK_L1198F
                 "2:29443613_C>T", #ALK_G1202R
                 "2:29443631_G>T", #ALK_L1196M
                 "2:29443610_C>T", #ALK_D1203N
                 "2:29443589_C>T", #ALK_E1210K
                 "2:29432682_C>G", #ALK_G1269A
                 "2:29445469_G>C", #ALK_L1122V
                 "2:29443600_G>C", #ALK_S1206C
                 "7:55249092_G>C", #EGFR_C797S
                 "7:55249071_C>T", #EGFR_T790M
                 "7:55259515_T>G") #EGFR_L858R


# Define the UPIs
upis <- c("MGH086-1", "MGH086-3", "MGH086-00", "MGH086-2", "MGH086-4", 
          "MGH086-5", "MGH086-0", "MGH086-000")
          
#Use the upis of the other three patients to produce the same file and change the file
#name output at the end of this block to the correct patient identifier because
#this is just for MGH086 added variants

#upis <- c("MGH808-A", "MGH808-E", "MGH808-F")

#upis <- c("MGH953-1", "MGH953-2-CL", "MGH953-3-CL", "MGH953-4-CL", "MGH953-6")

#upis <- c("MGH10032-AS-mediastinal", "MGH10032-LL-supraclavicular-LN", "MGH10032-R-hilar-LN", "MGH10032-R-paratracheal-LN", "MGH10032-RLL", "MGH10032-pre-treatment")


# Define the variant IDs and UPIs (as before)

# Initialize an empty data frame to store the results
results_df <- data.frame()

# Nested loop over variant IDs and UPIs (as before)
for (variant_id in variant_ids) {
  for (upi in upis) {
    # Call the hypothetical_source_prob function
    result <- hypothetical_source_prob(cesa, upi, variant_id)
    
    # Append the result to the data frame
    results_df <- rbind(results_df, result)
  }
}

# Write the results to a TSV file
write.table(results_df, file = "temp.tsv", sep = "\t", row.names = FALSE, quote = FALSE)

# Read the TSV file into a data frame
hyp_prob <- fread("temp.tsv")

# Create the mapping from variant IDs to gene names
name_mapping <- setNames(c("ALK_D1203N", "ALK_E1210K", "ALK_G1202R", "ALK_G1269A", 
                           "ALK_L1122V", "ALK_L1196M", "ALK_S1206C", "EGFR_C797S", 
                           "EGFR_T790M", "EGFR_L858R", "ALK_L1044F", "ALK_I1171N", "ALK_L1198F"),
                         c("2:29443610_C>T", "2:29443589_C>T", "2:29443613_C>T", 
                           "2:29432682_C>G", "2:29445469_G>C", "2:29443631_G>T", 
                           "2:29443600_G>C", "7:55249092_G>C", "7:55249071_C>T", 
                           "7:55259515_T>G", "2:29448369_G>A", "2:29445213_A>T","2:29443625_G>A"))

                           
# Ensure 'hyp_prob' is a data.table
setDT(hyp_prob)

# Rename the first column to 'variant_id' if it's not already named
setnames(hyp_prob, old = names(hyp_prob)[1], new = "variant_id")

# Use the name_mapping vector to replace the variant_id with gene names
hyp_prob[, variant_id := name_mapping[variant_id]]

# Now read the mutation rates data
mutation_rates <- fread("Mutation_rates_per_variant_named.tsv")

# Reshape mutation_rates from wide to long format
long_mutation_rates <- melt(mutation_rates, id.vars = "Unique_Patient_Identifier", 
                            variable.name = "variant_id", value.name = "VAR_MUTRATE")

# Merge hyp_prob with long_mutation_rates on 'Unique_Patient_Identifier' and 'variant_id'
merged_data <- merge(hyp_prob, long_mutation_rates, by = c("Unique_Patient_Identifier", "variant_id"), all.x = TRUE)


# Remove columns that contain only zero
merged_data <- merged_data[, sapply(merged_data, function(col) !all(col == 0)), with = FALSE]

# Move 'variant_id' column to the first position
setcolorder(merged_data, c("variant_id", setdiff(names(merged_data), "variant_id")))

# Write the merged data to a new TSV file
fwrite(merged_data, file="hyp_prob_MGH086_mutrates.tsv", sep="\t", quote=FALSE)

file.remove("temp.tsv")

# results now contains the hypothetical probabilities for each variant and UPI combination and the variant mutation rates


______________________________________________________________________________
______________________________________________________________________________

#Figure 1 code

#First create the file Figure1.tsv

# Step 1: Rename a new df Figure1 <- source_probabilities #assuming the data frame source_probabilities is stil existing in the global env
Figure1 <- source_probabilities

# Step 2: Concatenate the first 2 columns of the Figure1 df with a "_" between and name the new column Unique_Patient_Identifier
Figure1$Unique_Patient_Identifier <- paste(Figure1[[1]], Figure1[[2]], sep = "_")

# Explicitly remove the "variant_id" column if it exists
Figure1 <- subset(Figure1, select = -c(variant_id))

# Prepare the custom order of IDs for Unique_Patient_Identifier
ordered_ids <- c(
	"ALK_D1203N_MGH086-5",
	"ALK_E1210K_MGH086-5",
	"ALK_L1122V_MGH086-2",
	"ALK_E1210K_MGH086-0",
	"ALK_E1210K_MGH086-2",
	"ALK_S1206C_MGH086-0",
	"ALK_G1269A_MGH086-5",
	"ALK_D1203N_MGH086-1",
	"ALK_E1210K_MGH086-1",
	"ALK_L1196M_MGH953-6",
	"ALK_G1202R_MGH953-6
	"ALK_E1210K_MGH086-00",
	"ALK_D1203N_MGH086-3",
	"ALK_E1210K_MGH086-3",
	"ALK_D1203N_MGH086-4",
	"ALK_E1210K_MGH086-4",
	"EGFR_T790M_MGH808-E",
	"EGFR_L858R_MGH808-E",
	"EGFR_T790M_MGH10032-R-hilar-LN",
	"EGFR_L858R_MGH10032-R-hilar-LN",
	"EGFR_C797S_MGH10032-R-hilar-LN",
	"EGFR_L858R_MGH808-A",
	"ALK_G1202R_MGH953-4-CL",
	"EGFR_T790M_MGH808-F",
	"EGFR_L858R_MGH808-F",
	"EGFR_C797S_MGH808-F",
	"EGFR_L858R_MGH10032-pre-treatment",
	"EGFR_T790M_MGH10032-AS-mediastinal",
	"EGFR_L858R_MGH10032-AS-mediastinal",
	"EGFR_C797S_MGH10032-AS-mediastinal",
	"EGFR_C797S_MGH10032-LL-supraclavicular-LN",
	"EGFR_C797S_MGH10032-R-paratracheal-LN",
	"EGFR_C797S_MGH10032-RLL",
	"EGFR_L858R_MGH10032-LL-supraclavicular-LN",
	"EGFR_L858R_MGH10032-R-paratracheal-LN",
	"EGFR_L858R_MGH10032-RLL",
	"EGFR_T790M_MGH10032-LL-supraclavicular-LN",
	"EGFR_T790M_MGH10032-R-paratracheal-LN",
	"EGFR_T790M_MGH10032-RLL"
)

# Step 3: Reorder the lines of the df using the custom order for Unique_Patient_Identifier
Figure1$Unique_Patient_Identifier <- factor(Figure1$Unique_Patient_Identifier, levels = ordered_ids)
Figure1 <- Figure1[order(Figure1$Unique_Patient_Identifier), ]

# Step 4: Save the df as Figure1.tsv, ensuring variant_id is excluded
write.table(Figure1, "Figure1.tsv", sep = "\t", row.names = FALSE, quote = FALSE, col.names = TRUE)



# Load libraries
library(tidyverse)
library(ggplot2)
library(dplyr)

# Load the data
#In processed data folder
data <- read.delim('Figure1.tsv', sep = '\t')

# do it with an ordered list instead
# Your ordered list of Unique Patient Identifiers
ordered_ids2 <- rev(c(
	"ALK_D1203N_MGH086-5",
	"ALK_E1210K_MGH086-5",
	"ALK_L1122V_MGH086-2",
	"ALK_E1210K_MGH086-0",
	"ALK_E1210K_MGH086-2",
	"ALK_S1206C_MGH086-0",
	"ALK_G1269A_MGH086-5",
	"ALK_D1203N_MGH086-1",
	"ALK_E1210K_MGH086-1",
	"ALK_L1196M_MGH953-6",
	"ALK_G1202R_MGH953-6",
	"ALK_E1210K_MGH086-00",
	"ALK_D1203N_MGH086-3",
	"ALK_E1210K_MGH086-3",
	"ALK_D1203N_MGH086-4",
	"ALK_E1210K_MGH086-4",
	"EGFR_T790M_MGH808-E",
	"EGFR_L858R_MGH808-E",
	"EGFR_T790M_MGH10032-R-hilar-LN",
	"EGFR_L858R_MGH10032-R-hilar-LN",
	"EGFR_C797S_MGH10032-R-hilar-LN",
	"EGFR_L858R_MGH808-A",
	"ALK_G1202R_MGH953-4-CL",
	"EGFR_T790M_MGH808-F",
	"EGFR_L858R_MGH808-F",
	"EGFR_C797S_MGH808-F",
	"EGFR_L858R_MGH10032-pre-treatment",
	"EGFR_T790M_MGH10032-AS-mediastinal",
	"EGFR_L858R_MGH10032-AS-mediastinal",
	"EGFR_C797S_MGH10032-AS-mediastinal",
	"EGFR_C797S_MGH10032-LL-supraclavicular-LN",
	"EGFR_C797S_MGH10032-R-paratracheal-LN",
	"EGFR_C797S_MGH10032-RLL",
	"EGFR_L858R_MGH10032-LL-supraclavicular-LN",
	"EGFR_L858R_MGH10032-R-paratracheal-LN",
	"EGFR_L858R_MGH10032-RLL",
	"EGFR_T790M_MGH10032-LL-supraclavicular-LN",
	"EGFR_T790M_MGH10032-R-paratracheal-LN",
	"EGFR_T790M_MGH10032-RLL"
))
  
# Reshape the data into a long format
data_long <- data %>% pivot_longer(cols = -Unique_Patient_Identifier, 
                                   names_to = "signature", 
                                   values_to = "weight")

# Convert Unique_Patient_Identifier to a factor with the specified order
data_long$Unique_Patient_Identifier <- factor(data_long$Unique_Patient_Identifier, levels = ordered_ids2)

# Signature to process mapping
signature_process_mapping <- list(
  SBS1 = 'Deamination with age, clock-like (1)',
  SBS5 = 'Unknown, clock-like (5)',
  SBS13 = 'APOBEC (2,13)',
  SBS2 = 'APOBEC (2,13)',
  SBS3 = 'Defective homologous recombination (3)',
  SBS4 = 'Tobacco (4,29)',
  SBS7a = 'UV light (7a–d,38)',
  SBS7b = 'UV light (7a–d,38)',
  SBS7c = 'UV light (7a–d,38)',
  SBS7d = 'UV light (7a–d,38)',
  SBS38 = 'UV light (7a–d,38)',
  SBS11 = 'Prior treatment (11,31,32,35)',
  SBS22 = 'Mutagenic chemical exposure (22,24,42,88)',
  SBS24 = 'Mutagenic chemical exposure (22,24,42,88)',
  SBS42 = 'Mutagenic chemical exposure (22,24,42,88)',
  SBS88 = 'Mutagenic chemical exposure (22,24,42,88)',
  SBS16 = 'Alcohol-associated (16)',
  SBS29 = 'Tobacco (4,29)',
  SBS31 = 'Prior treatment (11,31,32,35)',
  SBS32 = 'Prior treatment (11,31,32,35)',
  SBS35 = 'Prior treatment (11,31,32,35)'
  # Add any additional signatures as needed
)

# Apply the mapping to the data
data_long$signature_process <- map_chr(data_long$signature, ~signature_process_mapping[[.]] %||% 'Non-actionable and unknown signatures')

#Aggregate the Weights: Then, aggregate the weights by Unique_Patient_Identifier and process.
#data_aggregated <- data_long %>%
#  group_by(Unique_Patient_Identifier, signature_process) %>%
#  summarise(weight = sum(weight))


# Filter out rows with weight = 0
data_long <- data_long %>% filter(weight > 0)

# Define the color palette
colors <- c(
  'APOBEC (2,13)' = '#8659a2',
  'Defective homologous recombination (3)' = '#ed0082',
  'Tobacco (4,29)' = '#a07627',
  'Unknown, clock-like (5)' = '#9d948d',
  'Deamination with age, clock-like (1)' = '#6a615c',
  'Non-actionable and unknown signatures' = 'black',
  'UV light (7a–d,38)' = '#ffffb3',
  'Prior treatment (11,31,32,35)' = '#80b1d3',
  'Mutagenic chemical exposure (22,24,42,88)' = '#fdb462',
  'Alcohol-associated (16)' = '#b3de69'
)

# Ensure the order of 'signature_process' matches the color vector
#data_long$signature_process <- factor(data_long$signature_process, levels = names(colors))
# Reverse the order of the levels of 'signature_process' to match the reversed 'colors' vector
data_long$signature_process <- factor(data_long$signature_process, levels = rev(names(colors)))

# Create and save the plots
unique_signature_processes <- unique(data_long$signature_process)

# Generate a single plot with all signature processes
p <- ggplot(data_long, aes(x = Unique_Patient_Identifier, y = weight, fill = signature_process)) +
    geom_bar(stat = "identity", color = "black", size = 0.75) + # Add black outline to bars
    scale_fill_manual(values = colors) + # Use predefined colors for each signature process
    labs(title = "Mutation Signature Processes", x = "Unique Patient Identifier", y = "Weight") +
    coord_flip() +
    theme_minimal()


#another p



# Save the plot as PDF and PNG
ggsave("all_signature_processes_plot.pdf", plot = p, width = 10, height = 7, dpi = 300)
ggsave("all_signature_processes_plot.png", plot = p, width = 10, height = 7, dpi = 300)


_______________________________________________________________________________
_______________________________________________________________________________

#Time gained with R calculation

# Load necessary libraries
library(readr)

#Create the files for time gain calculation for the observed variants
# Step 1: Load the file
data <- read.table("Source_probabilities_with_mutrates.tsv", header = TRUE, sep = "\t", stringsAsFactors = FALSE)

# Step 2: Create the new columns
data <- data %>%
  mutate(
    SBS2_SBS13 = SBS2 + SBS13,
    APOBEC_RATE = `SBS2_SBS13` * VAR_MUTRATE,
    SUPRESSED_RATE = VAR_MUTRATE - APOBEC_RATE
  )

# Define the groups of patient identifiers
patient_groups <- list(
  MGH086 = c("MGH086-1", "MGH086-3", "MGH086-4", "MGH086-5", "MGH086-0", "MGH086-00", 
             "MGH086-1", "MGH086-2", "MGH086-3", "MGH086-4", "MGH086-5", "MGH086-5", 
             "MGH086-2", "MGH086-0"),
  MGH953 = c("MGH953-4-CL", "MGH953-6", "MGH953-6"),
  MGH10032 = c("MGH10032-AS-mediastinal", "MGH10032-LL-supraclavicular-LN", "MGH10032-R-hilar-LN", "MGH10032-R-paratracheal-LN", "MGH10032-RLL", "MGH10032-AS-mediastinal", "MGH10032-LL-supraclavicular-LN", "MGH10032-R-hilar-LN", "MGH10032-R-paratracheal-LN", "MGH10032-RLL"),
  MGH808 = c("MGH808-E", "MGH808-F", "MGH808-F")
)

# Step 3: Split the file and save separate files
for (group_name in names(patient_groups)) {
  group_ids <- patient_groups[[group_name]]
  group_data <- filter(data, Unique_Patient_Identifier %in% group_ids)
  
  # Select only the necessary columns
  group_data <- group_data %>%
    select(Unique_Patient_Identifier, `SBS2_SBS13`, VAR_MUTRATE, APOBEC_RATE, SUPRESSED_RATE)
  
  # Write to a file
  write.table(group_data, paste0(group_name, "_time_calc.txt"), sep = "\t", row.names = FALSE, quote = FALSE)
}

____________________________________________________________________________

#Do the same for the MGH086 patient with the added variants (or for the other three patients)

# Step 1: Load the file
data <- read.table("hyp_prob_MGH10032_mutrates.tsv", header = TRUE, sep = "\t", stringsAsFactors = FALSE)

# Step 2: Create the new columns with valid variable names
data <- data %>%
  mutate(
    SBS2_SBS13 = SBS2 + SBS13,
    APOBEC_RATE = SBS2_SBS13 * VAR_MUTRATE,
    SUPRESSED_RATE = VAR_MUTRATE - APOBEC_RATE
  )

# Since all IDs are for MGH086, we don't need to split the data, just select the necessary columns
final_data <- data %>%
  select(Unique_Patient_Identifier, SBS2_SBS13, VAR_MUTRATE, APOBEC_RATE, SUPRESSED_RATE)

# Step 3: Save the data to a new file
write.table(final_data, "MGH10032_with_added_variants_time_calc.txt", sep = "\t", row.names = FALSE, quote = FALSE)

__________________________________________________________________________

#Calculation part equation 1

library(readr)

# Read the tab-delimited file (change the name each time and re-run the code)
data <- read_delim("MGH10032_time_calc.txt", delim = "\t")

# Calculate the estimated time gained for each row
data$estimated_time_gained <- ((1 / data$SUPRESSED_RATE) - (1 / data$VAR_MUTRATE)) / (1 / data$VAR_MUTRATE) * 100

# Calculate the sum for all rows
sum_suppressed_rate <- sum(data$SUPRESSED_RATE, na.rm = TRUE)
sum_variant_rate <- sum(data$VAR_MUTRATE, na.rm = TRUE)
total_estimated_time_gained <- ((1 / sum_suppressed_rate) - (1 / sum_variant_rate)) / (1 / sum_variant_rate) * 100


# Output the results for each row
#write_csv(data, "Estimated_Time_Gained_Per_Row_MGH086_calc.csv")

# Print the total estimated time gained for all rows
print(paste("Total estimated time gained for all variants:", total_estimated_time_gained))

__________________________________________________________________________
#Simpler version of equation 1 code for time gained COMPARING Suppressed with Variant rates, change the file name each time

library(readr)

# Load data
data <- read_delim("MGH086_with_added_variants_time_calc.txt", delim = "\t")

# Calculate Ts and Tu
Ts <- 1 / sum(data$SUPRESSED_RATE, na.rm = TRUE)
Tv <- 1 / sum(data$VAR_MUTRATE, na.rm = TRUE)

# Calculate the total estimated time gained as a percentage
total_estimated_time_gained <- ((Ts / Tv) - 1) * 100


print(Ts)
print(Tv)


# Print the total estimated time gained for all rows
print(paste("Total estimated time gained for all variants:", total_estimated_time_gained))


#PLOTS FOR Ts and Tv

#Plot Tv and Ts for each patient for observed and with all variants

library(tibble)
library(ggplot2)
library(viridis)
library(dplyr)

# Creating the data frame
data <- tribble(
  ~Patient, ~Condition, ~Type, ~Value,
  "MGH086", "Observed", "Tv", 2522.905,
  "MGH086", "Observed", "Ts", 12634.65,
  "MGH086", "All Variants", "Tv", 1168.657,
  "MGH086", "All Variants", "Ts", 7110.2,
  "MGH808", "Observed", "Tv", 28752.81,
  "MGH808", "Observed", "Ts", 35418.32,
  "MGH808", "All Variants", "Tv", 5339.864,
  "MGH808", "All Variants", "Ts", 9624.114,
  "MGH953", "Observed", "Tv", 177553.5,
  "MGH953", "Observed", "Ts", 331083.2,
  "MGH953", "All Variants", "Tv", 2632.301,
  "MGH953", "All Variants", "Ts", 7830.209,
  "MGH10032", "Observed", "Tv", 17354.89,
  "MGH10032", "Observed", "Ts", 18310.76,
  "MGH10032", "All Variants", "Tv", 3313.055,
  "MGH10032", "All Variants", "Ts", 4037.534 
)

#ggplot(data, aes(x = Patient, y = Value, fill = Type)) +
#  geom_bar(stat = "identity", position = "dodge") +
#  facet_wrap(~Condition, scales = "free", ncol = 1) +
#  coord_flip() +
#  theme_minimal() +
#  labs(y = "Value", x = "Patient",
#       title = "Ts and Tu Values for Each Patient",
#       subtitle = "Comparing Observed and All Variants Conditions",
#       fill = "Type") +
#  scale_fill_brewer(palette = "Set1")

# Create an empty list to store plots
plot_list <- list()

# Iterate over each patient to create a plot for each
unique_patients <- unique(data$Patient)

for(patient_id in unique_patients) {
  # Filter data for the current patient
  patient_data <- filter(data, Patient == patient_id)
  
# Ensure 'Type' is a factor and set the levels explicitly to order 'Tu' before 'Tv'
patient_data$Type <- factor(patient_data$Type, levels = c("Tv", "Tu"))
  
  # Create the plot for the current patient
p <- ggplot(patient_data, aes(x = Condition, y = Value, fill = Type)) +
  geom_bar(stat = "identity", position = "dodge", width = 0.45, color = "black") +
  coord_flip() +
  theme_minimal() +
  labs(y = "Value", x = "Condition",
       title = paste("Tu and Tv Values:", patient_id),
       subtitle = "Comparing Observed and All Variants",
       fill = "Type") +
  scale_fill_manual(values = c("Tu" = "#8659a2", "Tv" = "black"), name = "Type") +
  theme(text = element_text(color = "black"), # Set all text color to black
        axis.title = element_text(color = "black"), # Ensure axis titles are black
        axis.text = element_text(color = "black"), # Ensure axis texts are black
        legend.text = element_text(color = "black"), # Ensure legend text is black
        legend.title = element_text(color = "black"), # Ensure legend title is black
        plot.title = element_text(color = "black"), # Ensure plot title is black
        plot.subtitle = element_text(color = "black")) # Ensure plot subtitle is black

# Add the plot to the list
plot_list[[patient_id]] <- p
}

# Iterate over the list to save each plot as a PDF
for(patient_id in names(plot_list)) {
  pdf_file <- paste0(patient_id, "_Tu_Tv_plot.pdf")
  ggsave(pdf_file, plot = plot_list[[patient_id]], device = "pdf", width = 7, height = 2)
}

